src/storage/lock_manager/lock_wait_context.rs

```diff
@@ -0,0 +1,254 @@
+        start_ts: TimeStamp,
+        for_update_ts: TimeStamp,
+        first_batch_pr: ProcessResult,
+        cb: StorageCallback,
+        allow_lock_with_conflict: bool,
+    ) -> Self {
+        let inner = LockWaitContextInner {
+            pr: first_batch_pr,
+            cb,
+            lock_wait_token,
+        };
+        Self {
+            shared_states: Arc::new(LockWaitContextSharedState {
+                ctx_inner: Mutex::new(Some(inner)),
+                finished: AtomicBool::new(false),
+            }),
+            lock_manager,
+            allow_lock_with_conflict,
+            start_ts,
+            for_update_ts,
+        }
+    }
+
+    pub fn get_shared_states(&self) -> &Arc<LockWaitContextSharedState> {
+        &self.shared_states
+    }
+
+    /// Get the callback that should be invoked when finishes executing the
+    /// scheduler command that issued the lock-waiting.
+    ///
+    /// When we support partially finishing a pessimistic lock request (i.e.
+    /// when acquiring lock multiple keys in one single request, allowing
+    /// some keys to be locked successfully while the others are blocked or
+    /// failed), this will be useful for handling the result of the first
+    /// write batch. But currently, the first write batch of a lock-waiting
+    /// request is always empty, so the callback is just noop.
+    pub fn get_callback_for_first_write_batch(&self) -> StorageCallback {
+        StorageCallback::Boolean(Box::new(|res| {
+            res.unwrap();
+        }))
+    }
+
+    /// Get the callback that should be called when the request is woken up on a
+    /// key.
+    pub fn get_callback_for_blocked_key(&self) -> PessimisticLockKeyCallback {
+        let ctx = self.clone();
+        Box::new(move |res| {
+            ctx.finish_request(res);
+        })
+    }
+
+    /// Get the callback that's used to cancel a lock-waiting request. Usually
+    /// called by
+    /// [`WaiterManager`](crate::server::lock_manager::WaiterManager) due to
+    /// timeout.
+    pub fn get_callback_for_cancellation(&self) -> impl FnOnce(StorageError) {
+        let ctx = self.clone();
+        move |e| {
+            ctx.finish_request(Err(e.into()));
+        }
+    }
+
+    fn finish_request(&self, result: Result<PessimisticLockRes, SharedError>) {
+        let ctx_inner = if let Some(inner) = self.shared_states.ctx_inner.lock().take() {
+            inner
+        } else {
+            debug!("double invoking of finish_request of LockWaitContext";
+                "start_ts" => self.start_ts,
+                "for_update_ts" => self.for_update_ts
+            );
+            return;
+        };
+
+        self.shared_states.finished.store(true, Ordering::Release);
+
+        // TODO: Uncomment this after the corresponding change of `LockManager` is done.
+        // self.lock_manager
+        //     .remove_lock_wait(ctx_inner.lock_wait_token);
+
+        if !self.allow_lock_with_conflict {
```

Comment:
What does it stand for when self.allow_lock_with_conflict is false ?

Reply:
For the new mode, the pessimistic lock request would still be processed, while in the legacy mode the write conflict error would be returned to the client directly.

Reply:
I'll add more comments to the field in `PessimisticLockParameters`.